library-best-practices/JLBP-4.md

```diff
@@ -0,0 +1,44 @@
+[JLBP-4] Avoid dependencies on unstable libraries and features
+--------------------------------------------------------------
+
+- Unstable libraries are defined as libraries that allow for breaking changes to
+  their Public API within the same major version. For libraries following
+  semver, this means libraries with a 0.x.y version. (See [JLBP-3](JLBP-3.md)
+  for more details on the recommendations for following semver.)
+- Unstable features are defined as features that are not part of the stable
+  Public API of a stable library, often marked with annotations like `@Beta`.
+  (See [JLBP-3](JLBP-3.md) for more details on the recommendations for
+  annotating unstable features.)
+- If your library depends on an unstable library or feature, and that feature
+  experiences a breaking change between versions, your library will be locked to
+  a specific version of your dependency.
+  - If you expose the unstable feature on your library's surface, then your
+    library's current major version will be permanently locked to the version
+    you initially exposed, and you won't be able to upgrade without making a
+    breaking change to your users.
+  - If you only use the unstable feature in your implementation, then each minor
+    or patch version of your library will require a very specific version of
+    your dependency, and it will be unsafe for your users to upgrade your
+    library on its own, creating opportunities for hard-to-diagnose runtime
+    conflicts for users.
+- Given the consequences of depending unstable features in dependencies, avoid
+  doing so.
+  - Depending on unstable features between submodules of a single library is
+    considered acceptable, given there is a way for users to ensure they are
+    using compatible versions of the submodules together easily, e.g. by having
+    the library use the same version for all submodules, or by providing a BOM.
+- Additionally, avoid depending on libraries that implement non-finalized JSRs,
+  even if the library otherwise meets the criteria for stability.
```

Comment:
I'm not sure what you mean that a JSR "otherwise meets the criteria for stability." It seems like we haven't given any criteria for stability of a JSR, since it doesn't follow semantic versioning. Hmm, or maybe some JSRs do nowadays? But my impression is that JCache wasn't 1.0 until it was finalized -- possibly it didn't even have a version number until then? -- so it would have failed the semver criteria.

I guess I don't know much about how JSRs are usually versioned. I know there's a jsr305 "version 3," but I don't know that it removed anything, so possibly it's not following semver at all? Maybe the problem is that someone slapped 1.0 on jsr305 even though it wasn't finalized?

But this sounds kind of like we're just saying: "If it doesn't follow semver, then you still need to consider stability. Some JSRs use semver; JSR305 is weird because someone called in 1.0 even though it wasn't."

Sorry, I'm not sure what actionable comment I'm trying to make here.

- Maybe, if the weird thing is that someone called JSR305 "1.0," we should say that (except that it doesn't really cause stability problems per se, just the module weirdness, so maybe we don't even need to).
- If people have been doing this kind of thing with other JSRs, then maybe there's a more general point. Do we know whether they have been?
- Maybe we should call out that unreleased JSRs are an example of something that might not even have a version number, and that's another smell that something might be unstable?

Reply:
JSRs just use status labels AFAICT: https://jcp.org/en/introduction/glossary#status ; the status labels are what I'm trying to refer to. Essentially: even if the *implementation* of a JSR is 1.x, if the JSR itself is in status Dormant or Active (instead of Final), then avoid depending on it because the specification that it's based on can change (meaning the implementation would need to bump to 2.x or whatever).

Reply:
I will try to add more clarification to distinguish JSR status (which doesn't use semver) from JSR implementations (which would use semver).